presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveFilterPushdownLogicalOptimizer.java

```diff
@@ -0,0 +1,527 @@
+
+        TupleDomain<Subfield> domainPredicate = withColumnDomains(ImmutableMap.<Subfield, Domain>builder()
+                .putAll(hivePartitionResult.getUnenforcedConstraint()
+                        .transform(HiveFilterPushdownLogicalOptimizer::toSubfield)
+                        .getDomains()
+                        .orElse(ImmutableMap.of()))
+                .putAll(decomposedFilter.getTupleDomain()
+                        .transform(subfield -> !isEntireColumn(subfield) ? subfield : null)
+                        .getDomains()
+                        .orElse(ImmutableMap.of()))
+                .build());
+
+        Set<String> predicateColumnNames = new HashSet<>();
+        domainPredicate.getDomains().get().keySet().stream()
+                .map(Subfield::getRootName)
+                .forEach(predicateColumnNames::add);
+        // Include only columns referenced in the optimized expression. Although the expression is sent to the worker node
+        // unoptimized, the worker is expected to optimize the expression before executing.
+        extractAll(optimizedRemainingExpression).stream()
+                .map(VariableReferenceExpression::getName)
+                .forEach(predicateColumnNames::add);
+
+        Map<String, HiveColumnHandle> predicateColumns = predicateColumnNames.stream()
+                .map(columnHandles::get)
+                .map(HiveColumnHandle.class::cast)
+                .collect(toImmutableMap(HiveColumnHandle::getName, Functions.identity()));
+
+        SchemaTableName tableName = ((HiveTableHandle) tableHandle).getSchemaTableName();
+        return new ConnectorPushdownFilterResult(
+                metadata.getTableLayout(
+                        session,
+                        new HiveTableLayoutHandle(
+                                tableName,
+                                hivePartitionResult.getPartitionColumns(),
+                                // remove comments to optimize serialization costs
+                                pruneColumnComments(hivePartitionResult.getDataColumns()),
+                                hivePartitionResult.getTableParameters(),
+                                hivePartitionResult.getPartitions(),
+                                domainPredicate,
+                                decomposedFilter.getRemainingExpression(),
+                                predicateColumns,
+                                hivePartitionResult.getEnforcedConstraint(),
+                                hivePartitionResult.getBucketHandle(),
+                                hivePartitionResult.getBucketFilter(),
+                                true,
+                                createTableLayoutString(session, tableName, hivePartitionResult.getBucketHandle(), hivePartitionResult.getBucketFilter(), decomposedFilter.getRemainingExpression(), domainPredicate))),
+                TRUE_CONSTANT);
+    }
+
+    private abstract class Visitor
```

Comment:
We should combine all these visitors into one. No need to separate them apart.

Reply:
I'll check if that is doable. Using multiple visitors as the original code had multiple optimizer rules executed one after another. If both rules are executed mutually exclusively then this is possible. I'll check if it works.